File: Apna Collge/Stack/ValidParenthesis.java
import java.util.*;

public class ValidParenthesis {

    public static boolean isValid(String str){
        Stack<Character> st = new Stack<>();
        for(int i=0;i<str.length();i++){
            char c = str.charAt(i);
            if(c=='(' || c=='{' || c=='['){
                st.push(c);
            }
            else{
                // closing case
                // )))))))) check khali to nhi opeinh h hi nhi
                if(st.isEmpty()) return false;

                // pair form ho rha hai
                if((st.peek()=='(' && c==')')|| (st.peek()=='[' && c==']') || (st.peek()=='{' && c=='}') ){
                    st.pop();
                }
                else{
                    return false;
                }
            }
        }
        if(st.isEmpty()) return true;
        else return false;
    }
    public static void main(String[] args) {
        String str = "(){}{}{}";
        System.out.println(isValid(str));
    }    
}

File: Apna Collge/Stack/Practice Questions/validParenthesis.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define mod 1000000007
bool isValid(string str)
{
    int n = str.size();
    stack<char> st;
    for (int i = 0; i < n; i++)
    {
        char ch = str[i];

        if (ch == '(' || ch == '[' || ch == '{')
        {
            st.push(ch);
        }
        else
        {
            if (st.empty())
            {
                return false;
            }

            // check if pair form ho rha hai
            if ((st.top() == '(' && ch == ')') || (st.top() == '[' && ch == ']') ||
                (st.top() == '{' && ch == '}'))
            {
                st.pop();
            }
            else
            {
                return false;
            }
        }
    }

    if (st.empty())
        return true;
    else
        return false;
}
int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    string str = "(){}[]";
    cout << isValid(str) << endl;
    return 0;
}
